crates/goose/src/security/patterns.rs

```diff
@@ -0,0 +1,632 @@
+                patterns.insert(threat.name, regex);
+            }
+        }
+        patterns
+    };
+}
+
+/// Pattern matcher for detecting security threats
+pub struct PatternMatcher {
+    patterns: &'static HashMap<&'static str, Regex>,
+}
+
+impl PatternMatcher {
+    pub fn new() -> Self {
+        Self {
+            patterns: &COMPILED_PATTERNS,
+        }
+    }
+
+    /// Scan text for security threat patterns
+    pub fn scan_text(&self, text: &str) -> Vec<PatternMatch> {
+        let mut matches = Vec::new();
+
+        for threat in THREAT_PATTERNS {
+            if let Some(regex) = self.patterns.get(threat.name) {
+                if regex.is_match(text) {
+                    // Find all matches to get position information
+                    for regex_match in regex.find_iter(text) {
+                        matches.push(PatternMatch {
+                            threat: threat.clone(),
+                            matched_text: regex_match.as_str().to_string(),
+                            start_pos: regex_match.start(),
+                            end_pos: regex_match.end(),
+                        });
+                    }
+                }
+            }
+        }
+
+        // Sort by risk level (critical first) and position
+        matches.sort_by(|a, b| match (&a.threat.risk_level, &b.threat.risk_level) {
+            (RiskLevel::Critical, RiskLevel::Critical) => a.start_pos.cmp(&b.start_pos),
+            (RiskLevel::Critical, _) => std::cmp::Ordering::Less,
+            (_, RiskLevel::Critical) => std::cmp::Ordering::Greater,
+            (RiskLevel::High, RiskLevel::High) => a.start_pos.cmp(&b.start_pos),
+            (RiskLevel::High, _) => std::cmp::Ordering::Less,
+            (_, RiskLevel::High) => std::cmp::Ordering::Greater,
+            _ => a.start_pos.cmp(&b.start_pos),
+        });
+
+        matches
+    }
+
+    /// Get the highest risk level from matches
+    pub fn get_max_risk_level(&self, matches: &[PatternMatch]) -> Option<RiskLevel> {
+        matches
+            .iter()
+            .map(|m| &m.threat.risk_level)
+            .max_by(|a, b| match (a, b) {
+                (RiskLevel::Critical, RiskLevel::Critical) => std::cmp::Ordering::Equal,
+                (RiskLevel::Critical, _) => std::cmp::Ordering::Greater,
+                (_, RiskLevel::Critical) => std::cmp::Ordering::Less,
+                (RiskLevel::High, RiskLevel::High) => std::cmp::Ordering::Equal,
+                (RiskLevel::High, _) => std::cmp::Ordering::Greater,
+                (_, RiskLevel::High) => std::cmp::Ordering::Less,
+                (RiskLevel::Medium, RiskLevel::Medium) => std::cmp::Ordering::Equal,
+                (RiskLevel::Medium, _) => std::cmp::Ordering::Greater,
+                (_, RiskLevel::Medium) => std::cmp::Ordering::Less,
+                (RiskLevel::Low, RiskLevel::Low) => std::cmp::Ordering::Equal,
+            })
+            .cloned()
+    }
```

Comment:
here and the function above, this hand sorting seems quaint. we should just make RiskLevel sortable